domain/domain.go

```diff
@@ -1328,7 +1325,7 @@ func (do *Domain) UpdateTableStatsLoop(ctx sessionctx.Context) error {
 	}
 	do.SetStatsUpdating(true)
 	do.wg.Run(func() { do.updateStatsWorker(ctx, owner) })
-	if RunAutoAnalyze {
+	if variable.RunAutoAnalyze.Load() {
```

Comment:
Changing the code here does not take effect actually, because `UpdateTableStatsLoop` is only called on the bootstrap stage of TiDB if you check the [code](https://github.com/pingcap/tidb/blob/ebb91121643c2a5e17c3c8af8f36b83d61688789/session/session.go/#L2939) in `BootstrapSession()`.

This is fine before because when it is changed in the configuration file, TiDB always needs to restart and the `BootstrapSession()` is called, but making it a sysvar means it the value can be changed dynamically, so:

* When it was `FALSE` and set to `TRUE`, the auto-analyze worker will not be invoked.
* When it was `TRUE` and set to `FALSE`, the auto-analyze worker will still be there.

You can easily verify it by, for example, adding a log in [`HandleAutoAnalyze()`](https://github.com/pingcap/tidb/blob/ebb91121643c2a5e17c3c8af8f36b83d61688789/statistics/handle/update.go/#L1037-L1038).

So my suggestion is removing the `if` here:

```
// always runs the loop
do.wg.Run(func() { do.autoAnalyzeWorker(owner) })
```
And moving `if` to [here](https://github.com/pingcap/tidb/blob/ebb91121643c2a5e17c3c8af8f36b83d61688789/domain/domain.go/#L1508) like:
```
		select {
		case <-analyzeTicker.C:
			if owner.IsOwner() && variable.RunAutoAnalyze.Load() {
				statsHandle.HandleAutoAnalyze(do.InfoSchema())
			}
		case <-do.exit:
			return
		}
```
So that the sysvar takes effect really.

@chrysan Please let me know if you have any comment on it, thanks!

Reply:
LGTM